pkg/ruler/remotequerier_test.go

```diff
@@ -117,6 +118,114 @@ func TestRemoteQuerier_QueryReq(t *testing.T) {
 
 }
 
+func TestRemoteQuerier_SendRequest(t *testing.T) {
+	const errMsg = "this is an error"
+	var (
+		successfulResponse = &httpgrpc.HTTPResponse{
+			Code: http.StatusOK,
+			Headers: []*httpgrpc.Header{
+				{Key: "Content-Type", Values: []string{"application/json"}},
+			},
+			Body: []byte(`{
+						"status": "success","data": {"resultType":"vector","result":[]}
+					}`),
+		}
+		erroneousResponse = &httpgrpc.HTTPResponse{
+			Code: http.StatusBadRequest,
+			Headers: []*httpgrpc.Header{
+				{Key: "Content-Type", Values: []string{"application/json"}},
+			},
+			Body: []byte("this is an error"),
+		}
+	)
+
+	tests := map[string]struct {
+		response        *httpgrpc.HTTPResponse
+		err             error
+		expectedError   error
+		expectedRetries bool
+	}{
+		"errors with code 5xx are retried": {
+			err:             httpgrpc.Errorf(http.StatusInternalServerError, errMsg),
+			expectedError:   httpgrpc.Errorf(http.StatusInternalServerError, errMsg),
+			expectedRetries: true,
+		},
+		"context.Canceled error is retried": {
+			err:             context.Canceled,
+			expectedError:   context.Canceled,
+			expectedRetries: true,
+		},
+		"gRPC context.Canceled error is retried": {
+			err:             status.Error(codes.Canceled, context.Canceled.Error()),
+			expectedError:   status.Error(codes.Canceled, context.Canceled.Error()),
+			expectedRetries: true,
+		},
+		"context.DeadlineExceeded error is retried": {
+			err:             context.DeadlineExceeded,
+			expectedError:   context.DeadlineExceeded,
+			expectedRetries: true,
+		},
+		"gRPC context.DeadlineExceeded error is retried": {
+			err:             status.Error(codes.DeadlineExceeded, context.DeadlineExceeded.Error()),
+			expectedError:   status.Error(codes.DeadlineExceeded, context.DeadlineExceeded.Error()),
+			expectedRetries: true,
+		},
+		"errors with code 4xx are not retried": {
+			err:             httpgrpc.Errorf(http.StatusBadRequest, errMsg),
+			expectedError:   httpgrpc.Errorf(http.StatusBadRequest, errMsg),
+			expectedRetries: false,
+		},
+		"non-erroneous responses with status code 4xx are not retried and are converted to errors": {
```

Comment:
[nit] I think calling them "non erroneous" is a bit misleading. I would simplify it.

```suggestion
		"responses with status code 4xx are not retried and are converted to errors": {
```